Parsimony/Parser.cs

```diff
@@ -0,0 +1,212 @@
+        }
+
+        /// <summary>
+        /// Parses a <typeparamref name="T"/> from <paramref name="input"/>.
+        /// </summary>
+        /// <param name="input">The tokens to parse.</param>
+        /// <returns>The <see cref="ParseResult{T}"/>.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="input"/> is <c>null</c>.
+        /// </exception>
+        public ParseResult<T> Parse(IEnumerable<string> input)
+        {
+            var tokens = (input ?? throw new ArgumentNullException(nameof(input))).ToList();
+
+            var options = new T();
+            var arguments = new List<string>(tokens.Count);
+            var result = new ParseResult<T>(options, arguments);
+            return Parse(result, tokens);
+        }
+
+        /// <summary>
+        /// Parses a <typeparamref name="T"/> from <paramref name="input"/>.
+        /// </summary>
+        /// <param name="input">The tokens to parse.</param>
+        /// <returns>The <see cref="ParseResult{T}"/>.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="input"/> is <c>null</c>.
+        /// </exception>
+        public ParseResult<T> Parse(params string[] input)
+        {
+            return Parse(input as IEnumerable<string>);
+        }
+
+        private ParseResult<T> Parse(ParseResult<T> result, List<string> input)
+        {
+            var state = new ParserState(result, input);
+            while (state.Input.Count > 0)
+                state = Parse(state);
+            return state.Result;
+        }
+
+        private ParserState Parse(ParserState state)
+        {
+            return ConsumeShortNameOption(state)
+                ?? ConsumeLongNameOption(state)
+                ?? ConsumeArgument(state)
+                ?? state;
+        }
+
+        private ParserState? ConsumeShortNameOption(ParserState state)
+        {
+            var options = state.Result.Options;
+            var arguments = state.Result.Arguments.ToList();
+            var input = state.Input;
+
+            if (input.Count == 0)
+                return null;
+
+            var match = _shortNameOptionPatern.Match(input[0]);
+            if (!match.Success)
+                return null;
+
+            var optionName = match.Groups["option"].Value[0];
+            var option = _options.FirstOrDefault(o => o.Spec.ShortName == optionName);
+            if (option == null)
+                // TODO: Custom exception
+                // TODO: Support for unknown options as arguments
+                throw new Exception($"Unknown argument '-{optionName}'");
+
+            // Consume the token
+            input = input.Skip(1).ToList();
+
+            var optionType = option.Spec.OptionType;
+            var typedOption = typeof(OptionSpec<,>).MakeGenericType(new[] { typeof(T), optionType });
+            var set = typedOption.GetMethod("Set", new[] { typeof(T), optionType });
```

Comment:
Why not just get the property with GetProperty?

Reply:
Oh this is based off of a selector? Why not just expose it not through reflection given that neither case really has any type safety

Reply:
This one is type safe because the Set is declared on the `OptionSpec<A>` but its impl is the field on `OptionSpec<A,B>` whose type params will be `{ typeof(T), optionType }`

Reply:
In order to get them into a collection together I made the `OptionSpec<A>` type to erase the type of the option which is why the `Set` method works the way it does. I'm open to something else, but nothing else gave me the same level of certainty about the type smatching

Reply:
I think that this might be solved by revisiting the object structure like we were talking about below

Reply:
agreed